elastic/kibana
# [Upgrade Assistant] Update "remove settings" action

## Summary

Fixes https://github.com/elastic/kibana/issues/124396

Elasticsearch recently added the ability to indicate any deprecated index and cluster settings that can be safely removed via the deprecation info API (related changes: https://github.com/elastic/elasticsearch/issues/83116, https://github.com/elastic/elasticsearch/pull/84145, https://github.com/elastic/elasticsearch/pull/84246).

The Upgrade Assistant can now use this information to auto-resolve deprecated index and cluster settings. Previously, UA only supported removing deprecated `translog` index settings, and this was determined by parsing a message string in the deprecation info API (brittle and error prone, see: https://github.com/elastic/kibana/pull/123599).

// Before - UA parsed message string from ES API to determine if an index setting can be removed
```
index_settings": {
  "test_index": [
    {
    "level": "warning",
    "message": "Translog retention settings are deprecated",
    "url": "https://ela.st/es-deprecation-7-translog-settings",
    "details": "Remove the translog retention settings: \"index.translog.retention.size\" and \"index.translog.retention.age\". The translog has not been used in peer recoveries with soft-deletes enabled since 7.0 and these settings have no effect.",
    "resolve_during_rolling_upgrade": false
    }
   ]
},
```

// After - UA can leverage the `_meta` property to determine whether an index setting can be removed. This action type is available for index *and* cluster settings.

```
index_settings": {
  "test_index": [
    {
    "level": "warning",
    "message": "Translog retention settings are deprecated",
    "url": "https://ela.st/es-deprecation-7-translog-settings",
    "details": "Remove the translog retention settings: \"index.translog.retention.size\" and \"index.translog.retention.age\". The translog has not been used in peer recoveries with soft-deletes enabled since 7.0 and these settings have no effect.",
    "resolve_during_rolling_upgrade": false,
    "_meta": {
       "actions": [
          "action_type": "remove_settings",
          "objects": ["index.translog.retention.size", "index.translog.retention.age"]
        ]
      }
    }
   ]
},
```

**Note on implementation**
The UX is largely the same for removing an index setting vs. cluster setting. However, different APIs are called and the copy is slightly different. I decided to create separate components to handle removing cluster settings rather than a single one with conditional logic. 

### How to test
To test, create some deprecated cluster settings, as well as an index with deprecated index settings. Verify you are able to remove the settings successfully via Upgrade Assistant.

```
// Test both persistent and transient settings
PUT /_cluster/settings
{
   "persistent": {
    "discovery.zen.commit_timeout": "10s"
  },
  "transient": {
    "discovery.zen.publish_timeout": "10s"
  }
}

```

```
PUT test_index
  {
    "settings": {
      "index.routing.allocation.include._tier": "data_warm",
      "index.routing.allocation.exclude._tier": "data_cold"
    }
  }
```

### Screenshots
// List of deprecations with "Remove settings" action
<img width="1422" alt="Screen Shot 2022-02-23 at 2 40 48 PM" src="https://user-images.githubusercontent.com/5226211/155400215-e64bd2b6-1920-490a-8dc8-47469e12fbfd.png">

// Remove cluster settings flyout
<img width="1075" alt="Screen Shot 2022-02-23 at 2 41 35 PM" src="https://user-images.githubusercontent.com/5226211/155400254-215f58c3-2529-4f4b-ae3a-baddcc6d628b.png">

// Error state for cluster settings flyout
<img width="529" alt="Screen Shot 2022-02-23 at 3 00 14 PM" src="https://user-images.githubusercontent.com/5226211/155400279-82a9c2f3-0dfb-4ed1-9e00-43e6a3471ca6.png">
